### ISSUE TITLE
Some questions with using your code

### ISSUE THREAD

**Issue by Iceland-Leo** [ZJCV/ZCls#7]
It's a nice work!! But I have a few questions might need you to answer: 
1. For the ZCLS version of “resnest50_fast_2s1x64d” and the Official version, the only difference is the FEATURE_DIMS obtained by backbone? I observed that the ZCLS version is 2048 and the official version is 1024. Are there any other differences?
2. The “resnest50_fast_2s1x64d” pretrained model you provided in Baidu Cloud Disk is ZCLS version? Could you provide an official version pretrained model in this [table](https://zcls.readthedocs.io/en/latest/benchmark-resnest/)?
3. The last and most important question,  if I want to use your pretraining model ("resnest50_fast_2s1x64d" for me) to train custom data, how do I set the learning rate to get the best performance? (Multi-GPU training)

**Comment by zjykzj** [comment]
> It's a nice work!! But I have a few questions might need you to answer:
> 
>     1. For the ZCLS version of “resnest50_fast_2s1x64d” and the Official version, the only difference is the FEATURE_DIMS obtained by backbone? I observed that the ZCLS version is 2048 and the official version is 1024. Are there any other differences?
> 
>     2. The “resnest50_fast_2s1x64d” pretrained model you provided in Baidu Cloud Disk is ZCLS version? Could you provide an official version pretrained model in this [table](https://zcls.readthedocs.io/en/latest/benchmark-resnest/)?
> 
>     3. The last and most important question,  if I want to use your pretraining model ("resnest50_fast_2s1x64d" for me) to train custom data, how do I set the learning rate to get the best performance? (Multi-GPU training)

hi @Iceland-Leo, 

* As far as I know, the official implementation of resnest has undergone many adjustments. I only refer to one of the implementations, so there may be some differences;
* you mentioned pretrained model in my docs is all from resnest official. I just convert it to my realization. From this point of view, it can be proved that I am consistent with the official implementation (maybe not latest)
* Finally, it's a pity that resnest is not used in my work, so it's difficult for me to give your best advice. You can check the paper for relevant information

Finally, I also welcome you to provide relevant training config. Have a nice day !!!

**Comment by Iceland-Leo** [comment]
OK, another question, the “resnest50_fast_2s1x64d” model doesn't seem to be faster than "resnest50", is this my implementation problem? The only difference seems to be the "fast_avg" parameter? Any other differences? (BTW, I migrated the model you implemented into my framework.)

**Comment by Iceland-Leo** [comment]
> OK, another question, the “resnest50_fast_2s1x64d” model doesn't seem to be faster than "resnest50", is this my implementation problem? The only difference seems to be the "fast_avg" parameter? Any other differences? (BTW, I migrated the model you implemented into my framework.)

In other words, the data given in your table shows that the inference speed difference between the "resnest50_fast_2s1x64d" model and the "ResNet50" model is not much different, but in fact the "resnest50_fast_2s1x64d" model that I migrated from yours is much slower.

**Comment by zjykzj** [comment]
> > OK, another question, the “resnest50_fast_2s1x64d” model doesn't seem to be faster than "resnest50", is this my implementation problem? The only difference seems to be the "fast_avg" parameter? Any other differences? (BTW, I migrated the model you implemented into my framework.)
> 
> In other words, the data given in your table shows that the inference speed difference between the "resnest50_fast_2s1x64d" model and the "ResNet50" model is not much different, but in fact the "resnest50_fast_2s1x64d" model that I migrated from yours is much slower.

check https://github.com/ZJCV/ZCls/blob/e6f14042bb7ed73667f42d36a8be9a69c872c43c/tools/benchmarks.py#L277

**Comment by Iceland-Leo** [comment]
yeah, I checked the files, especially the config file. But I only found the "fast" parameter setting in the code below. I followed this setup. I can see from your [table](https://zcls.readthedocs.io/en/latest/benchmark-resnest/) that "resnest50 fast" is only a little bit slower than "resnet50", but actually in my case "resnest50" is much slower, is there anything else I haven't taken into account?
> https://github.com/ZJCV/ZCls/blob/e6f14042bb7ed73667f42d36a8be9a69c872c43c/zcls/model/backbones/resnet/resnet_d_backbone.py#L270

**Comment by zjykzj** [comment]
> yeah, I checked the files, especially the config file. But I only found the "fast" parameter setting in the code below. I followed this setup. I can see from your [table](https://zcls.readthedocs.io/en/latest/benchmark-resnest/) that "resnest50 fast" is only a little bit slower than "resnet50", but actually in my case "resnest50" is much slower, is there anything else I haven't taken into account?
> 
> > https://github.com/ZJCV/ZCls/blob/e6f14042bb7ed73667f42d36a8be9a69c872c43c/zcls/model/backbones/resnet/resnet_d_backbone.py#L270

i just done it using my laptop, may be different computer config ?

**Comment by zjykzj** [comment]
> > yeah, I checked the files, especially the config file. But I only found the "fast" parameter setting in the code below. I followed this setup. I can see from your [table](https://zcls.readthedocs.io/en/latest/benchmark-resnest/) that "resnest50 fast" is only a little bit slower than "resnet50", but actually in my case "resnest50" is much slower, is there anything else I haven't taken into account?
> > > https://github.com/ZJCV/ZCls/blob/e6f14042bb7ed73667f42d36a8be9a69c872c43c/zcls/model/backbones/resnet/resnet_d_backbone.py#L270
> 
> i just done it using my laptop, may be different computer config ?

```
/home/zj/anaconda3/envs/py3.8/bin/python /home/zj/yoyo/ZCls/tools/benchmarks.py
##############################
/home/zj/anaconda3/envs/py3.8/lib/python3.8/site-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
/home/zj/anaconda3/envs/py3.8/lib/python3.8/site-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /pytorch/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
se_r50_zcls_imagenet **********
device: cpu
GFlops: 8.239G
Params Size: 107.086MB
compute cpu infer time
one process need 0.087s, model compute need: 0.087s
sknet50_zcls_imagenet **********
device: cpu
GFlops: 9.000G
Params Size: 104.885MB
compute cpu infer time
one process need 0.114s, model compute need: 0.111s
resnest50_fast_2s1x64d_zcls **********
device: cpu
GFlops: 8.719G
Params Size: 104.840MB
compute cpu infer time
one process need 0.091s, model compute need: 0.087s
resnest50_fast_2s1x64d_official **********
device: cpu
GFlops: 8.716G
Params Size: 104.840MB
compute cpu infer time
one process need 0.092s, model compute need: 0.087s
resnest50_zcls **********
device: cpu
GFlops: 10.805G
Params Size: 104.840MB
compute cpu infer time
one process need 0.107s, model compute need: 0.107s
resnest50_official **********
device: cpu
GFlops: 10.801G
Params Size: 104.840MB
compute cpu infer time
one process need 0.102s, model compute need: 0.099s
resnest101_zcls **********
device: cpu
GFlops: 20.496G
Params Size: 184.155MB
compute cpu infer time
one process need 0.186s, model compute need: 0.183s
resnest101_official **********
device: cpu
GFlops: 20.491G
Params Size: 184.155MB
compute cpu infer time
one process need 0.181s, model compute need: 0.178s
resnest200_zcls **********
device: cpu
GFlops: 34.991G
Params Size: 267.798MB
compute cpu infer time
one process need 0.341s, model compute need: 0.328s
resnest200_official **********
device: cpu
GFlops: 34.981G
Params Size: 267.798MB
compute cpu infer time
one process need 0.337s, model compute need: 0.324s
resnest269_zcls **********
device: cpu
GFlops: 45.083G
Params Size: 423.162MB
compute cpu infer time
one process need 0.445s, model compute need: 0.436s
resnest269_official **********
device: cpu
GFlops: 45.070G
Params Size: 423.162MB
compute cpu infer time
one process need 0.439s, model compute need: 0.436s

Process finished with exit code 0
```

**Comment by zjykzj** [comment]
```
/home/zj/anaconda3/envs/py3.8/bin/python /home/zj/yoyo/ZCls/tools/benchmarks.py
##############################
/home/zj/anaconda3/envs/py3.8/lib/python3.8/site-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
r18_torchvision **********
device: cpu
GFlops: 3.638G
Params Size: 44.592MB
/home/zj/anaconda3/envs/py3.8/lib/python3.8/site-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /pytorch/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
compute cpu infer time
one process need 0.031s, model compute need: 0.029s
r18_zcls **********
device: cpu
GFlops: 3.638G
Params Size: 44.592MB
compute cpu infer time
one process need 0.032s, model compute need: 0.031s
r34_torchvision **********
device: cpu
GFlops: 7.343G
Params Size: 83.152MB
compute cpu infer time
one process need 0.055s, model compute need: 0.051s
r34_zcls **********
device: cpu
GFlops: 7.343G
Params Size: 83.152MB
compute cpu infer time
one process need 0.057s, model compute need: 0.054s
r50_torchvision **********
device: cpu
GFlops: 8.223G
Params Size: 97.492MB
compute cpu infer time
one process need 0.077s, model compute need: 0.073s
r50_zcls **********
device: cpu
GFlops: 8.223G
Params Size: 97.492MB
compute cpu infer time
one process need 0.078s, model compute need: 0.075s
r101_torchvision **********
device: cpu
GFlops: 15.668G
Params Size: 169.942MB
compute cpu infer time
one process need 0.134s, model compute need: 0.133s
r101_zcls **********
device: cpu
GFlops: 15.668G
Params Size: 169.942MB
compute cpu infer time
one process need 0.131s, model compute need: 0.128s
r152_torchvision **********
device: cpu
GFlops: 23.118G
Params Size: 229.617MB
compute cpu infer time
one process need 0.194s, model compute need: 0.186s
r152_zcls **********
device: cpu
GFlops: 23.118G
Params Size: 229.617MB
compute cpu infer time
one process need 0.212s, model compute need: 0.221s
rxt50_32x4d_torchvision **********
device: cpu
GFlops: 8.519G
Params Size: 95.478MB
compute cpu infer time
one process need 0.115s, model compute need: 0.102s
rxt50_32x4d_zcls **********
device: cpu
GFlops: 8.519G
Params Size: 95.478MB
compute cpu infer time
one process need 0.100s, model compute need: 0.092s
rxt101_32x8d_torchvision **********
device: cpu
GFlops: 32.953G
Params Size: 338.712MB
compute cpu infer time
one process need 0.341s, model compute need: 0.342s
rxt101_32x8d_zcls **********
device: cpu
GFlops: 32.953G
Params Size: 338.712MB
compute cpu infer time
one process need 0.304s, model compute need: 0.308s

Process finished with exit code 0
```

**Comment by zjykzj** [comment]
above benchmark records is just using CPU. By the way, I recommonded you to use ONNXRUNTIME or OpenVINO: [ zjykzj/onnx](https://github.com/zjykzj/onnx), you can get more speed infer

**Comment by Iceland-Leo** [comment]
I'm running on a GPU, the same device should also get relative results. Maybe I have some implementation problems


发自我的iPhone


------------------ Original ------------------
From: zhujian ***@***.***&gt;
Date: Thu,Sep 9,2021 10:24 PM
To: ZJCV/ZCls ***@***.***&gt;
Cc: Liucheng ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [ZJCV/ZCls] Some questions with using your code (#7)





 
above benchmark records is just using CPU. By the way, I recommonded you to use ONNXRUNTIME or OpenVINO:  zjykzj/onnx, you can get more speed infer
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
